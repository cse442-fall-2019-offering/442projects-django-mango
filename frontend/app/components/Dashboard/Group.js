import React from 'react';
import PropTypes from 'prop-types';

import { withStyles } from '@material-ui/core/styles';

const styles = () => ({
  group: {
    borderRadius: '10px',
  },
  name: {
    textAlign: 'center',
    fontWeight: 'bold',
    color: '#000',
    paddingTop: '20px',
  },
  languages: {
    textAlign: 'center',
    color: '#ffff00',
    marginTop: '50px',
  },
  language: {
    marginTop: '-20px',
  },
});

const Group = props => {
  const { classes } = props;
  return (
    <div className={classes.group}>
      <div className={classes.name}>{props.group[1]}</div>
      <div className={classes.languages}>
        {props.group[2].map(language => (
          <div className={classes.language} key={language}>
            <p>{`${language}\n`}</p>
          </div>
        ))}
      </div>
    </div>
  );
};

Group.propTypes = {
  classes: PropTypes.object.isRequired,
  group: PropTypes.array.isRequired,
};

export default withStyles(styles)(Group);
