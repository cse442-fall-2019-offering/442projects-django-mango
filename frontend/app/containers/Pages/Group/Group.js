import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Prompt } from 'react-router';
import { BrowserRouter } from 'react-router-dom';
import { createStructuredSelector } from 'reselect';

import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';
import AccountCircle from '@material-ui/icons/AccountCircle';
import { withStyles } from '@material-ui/core/styles';
import firebase from 'firebase';

import {
  getGroup,
  updateGroup,
  joinGroup,
  leaveGroup,
} from 'my-actions/groupActions';
import Loading from 'my-components/Loading';
import ReactMediumEditor from 'my-components/Editor/ReactMediumEditor';
import Navbar from 'my-components/Navbar/Navbar';
import { makeSelectGroup } from 'my-selectors/groupSelectors';
import NotFoundPage from 'containers/NotFoundPage/Loadable';
import styles from './group-jss';

let currentUser;

function containsObject(obj, list) {
  for (let i = 0; i < list.length; i += 1) {
    if (list[i] === obj) {
      return true;
    }
  }
  return false;
}

class Group extends Component {
  state = {
    name: '',
    description: '',
    languages: [],
    members: [],
    changed: [],
    edit: false,
    loading: true,
    error: false,
    isAuthenticating: true,
  };

  static getDerivedStateFromProps(props) {
    if (props.group.length < 1) {
      return {
        loading: true,
      };
    }
    if (props.group.error) {
      return {
        loading: false,
        error: true,
      };
    }
    return {
      name: props.group.name,
      description: props.group.description,
      languages: props.group.languages,
      members: props.group.members,
      loading: false,
    };
  }

  componentDidMount() {
    firebase.initializeApp({
      apiKey: 'AIzaSyDyQlqoOHI2Af0dzNswbZ4T-B9qicu4ByU',
      authDomain: 'django-mango.firebaseapp.com',
    });

    this.authUser().then(
      () => {
        this.setState({ isAuthenticating: false });
      },
      () => {
        this.setState({ isAuthenticating: false });
        // alert(e);
      },
    );

    const { onGetGroup } = this.props;
    const payload = { groupId: this.props.match.params.groupId };
    onGetGroup(payload);
  }

  authUser() {
    return new Promise(function auth(resolve, reject) {
      firebase.auth().onAuthStateChanged(function authStateChanged(user) {
        if (user) {
          resolve(user);
          if (user.email.includes('@buffalo.edu')) {
            currentUser = user;
          } else {
            alert('You must sign-in with an @buffalo.edu email address.');
            firebase.auth().signOut();
            window.location.href = '/';
          }
        } else {
          reject(new Error('User not logged in'));
          window.location.href = '/';
        }
      });
    });
  }

  handleJoinGroup = () => {
    const { onJoinGroup } = this.props;
    const payload = { groupId: this.props.match.params.groupId };
    onJoinGroup(payload);
  };

  handleLeaveGroup = () => {
    const { onLeaveGroup } = this.props;
    const payload = { groupId: this.props.match.params.groupId };
    onLeaveGroup(payload);
  };

  handleUpdateGroup = () => {
    const { onUpdateGroup } = this.props;
    const payload = {
      groupId: this.props.match.params.groupId,
      name: this.state.name,
      description: this.state.description,
      languages: this.state.languages,
    };
    onUpdateGroup(payload);
    this.setState({
      changed: [],
    });
  };

  handleNameChange = event => {
    this.setState({
      name: event.target.value,
    });
    if (!containsObject('name', this.state.changed))
      this.setState(previousState => ({
        changed: [...previousState.changed, 'name'],
      }));
  };

  handleDescriptionChange = description => {
    this.setState({
      description,
    });
    if (!containsObject('description', this.state.changed))
      this.setState(previousState => ({
        changed: [...previousState.changed, 'description'],
      }));
  };

  handleLanguagesChange = languages => {
    this.setState({
      languages,
    });
    if (!containsObject('languages', this.state.changed))
      this.setState(previousState => ({
        changed: [...previousState.changed, 'languages'],
      }));
  };

  editGroup = () => {
    this.setState(previousState => ({
      edit: !previousState.edit,
    }));
  };

  render() {
    if (this.state.loading || this.state.isAuthenticating) {
      return <Loading />;
    }
    if (this.state.error) {
      return <NotFoundPage />;
    }
    const { classes } = this.props;
    return (
      <BrowserRouter>
        <div className={classes.root}>
          <Prompt
            when={this.state.changed.length > 0}
            message="You have unsaved changes, are you sure you want to leave?"
          />
          <Navbar email={currentUser.email} />
          <Grid container spacing={2} justify="center">
            <Grid item md={3} sm={10} xs={12}>
              <div className={classes.button}>
                <Button
                  variant="contained"
                  color="primary"
                  type="button"
                  onClick={this.handleJoinGroup}
                >
                  Join
                </Button>
              </div>
              <div className={classes.name}>{this.state.name}</div>
              <ReactMediumEditor
                className={classes.description}
                text={this.state.description}
                onChange={this.handleDescriptionChange}
                placeholder="Group Description"
              />
              <div className={classes.languages}>
                <div className={classes.languageTitle}>
                  <p>Languages</p>
                </div>
                {this.state.languages.map(language => (
                  <p key={language}>{`${language}\n`}</p>
                ))}
              </div>
              <div className={classes.members}>
                {this.state.members.map(member => (
                  <p key={member}>
                    <AccountCircle />
                    {` ${member}\n`}
                  </p>
                ))}
              </div>
            </Grid>
          </Grid>
        </div>
      </BrowserRouter>
    );
  }
}

Group.propTypes = {
  classes: PropTypes.object.isRequired,
  onGetGroup: PropTypes.func.isRequired,
  onUpdateGroup: PropTypes.func,
  onJoinGroup: PropTypes.func,
  onLeaveGroup: PropTypes.func,
  match: PropTypes.shape({
    params: PropTypes.shape({
      groupId: PropTypes.string.isRequired,
    }),
  }),
};

const mapStateToProps = createStructuredSelector({
  group: makeSelectGroup(),
});

const mapDispatchToProps = dispatch => ({
  onGetGroup: payload => dispatch(getGroup(payload)),
  onUpdateGroup: payload => dispatch(updateGroup(payload)),
  onJoinGroup: payload => dispatch(joinGroup(payload)),
  onLeaveGroup: payload => dispatch(leaveGroup(payload)),
});

const GroupMapped = connect(
  mapStateToProps,
  mapDispatchToProps,
)(Group);

export default withStyles(styles)(GroupMapped);
