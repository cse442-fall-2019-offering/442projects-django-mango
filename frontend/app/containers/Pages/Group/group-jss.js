const styles = theme => ({
  root: {
    margin: '0 auto',
    maxWidth: '65%',
    [theme.breakpoints.down('sm')]: {
      maxWidth: '100%',
    },
    maxHeight: '100%',
  },

  button: {
    position: 'relative',
    top: '170px',
    left: 20,
    width: 170,
    height: 40,
    paddingTop: '50px',
  },

  name: {
    position: 'relative',
    top: 100,
    left: 300,
    width: 500,
    fontSize: '50px',
    fontWeight: 'bold',
    textAlign: 'center',
  },

  description: {
    position: 'relative',
    top: 150,
    left: 300,
    width: 510,
    height: 400,
    zIndex: 1,
    backgroundColor: '#fff',
    color: 'black',
    borderRadius: 18,
    padding: 50,
    overflow: 'auto',
    boxShadow: '0px 4px 10px rgba(51, 50, 50, 0.15)',
  },

  languageTitle: {
    color: '#0000ff',
  },

  languages: {
    position: 'relative',
    bottom: 200,
    left: 50,
    width: 400,
    height: 500,
    fontSize: '20px',
  },

  languagesMenu: {
    position: 'relative',
    bottom: 180,
    width: 220,
    height: 500,
    fontSize: '15px',
  },

  members: {
    position: 'relative',
    bottom: 720,
    left: 940,
    width: 400,
    height: 500,
    fontSize: '20px',
  },

  contact: {
    position: 'relative',
    bottom: 1320,
    left: 740,
    width: 50,
    height: 50,
    zIndex: 5,
  },

  contactEdit: {
    position: 'relative',
    bottom: 1350,
    left: 740,
    width: 50,
    height: 50,
    zIndex: 5,
  },

  contactPopup: {
    height: 400,
    paddingTop: 20,
    paddingBottom: 20,
    paddingLeft: 40,
    paddingRight: 40,
    overflow: 'auto',
  },

  submitButton: {
    position: 'relative',
    bottom: 1430,
    left: 940,
    width: 170,
    height: 40,
    paddingTop: '50px',
  },

  close: {
    position: 'relative',
    bottom: 1385,
    right: 80,
    width: 50,
    height: 50,
    zIndex: 5,
  },

  edit: {
    position: 'relative',
    bottom: 1250,
    left: 785,
    width: 50,
    height: 50,
    zIndex: 5,
  },

  publicButton: {
    position: 'relative',
    bottom: 1407,
    left: 115,
  },
});

export default styles;
