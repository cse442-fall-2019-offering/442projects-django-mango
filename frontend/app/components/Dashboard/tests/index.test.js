import React from 'react';
import renderer from 'react-test-renderer';
import Group from '../Group';

test('Group Dashboard Component', () => {
  const component = renderer.create(
    <Group
      group={[
        '1E3r5g',
        'Django Mango',
        ['python', 'javascript'],
        ['test1@buffalo.edu', 'test2@buffalo.edu'],
      ]}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});
