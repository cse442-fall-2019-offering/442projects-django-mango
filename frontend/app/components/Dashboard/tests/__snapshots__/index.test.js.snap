// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Group Dashboard Component 1`] = `
<div
  className="Group-group-1"
>
  <div
    className="Group-name-2"
  >
    Django Mango
  </div>
  <div
    className="Group-languages-3"
  >
    <div
      className="Group-language-4"
    >
      <p>
        python

      </p>
    </div>
    <div
      className="Group-language-4"
    >
      <p>
        javascript

      </p>
    </div>
  </div>
</div>
`;
